Jenkinsfile

```diff
@@ -0,0 +1,71 @@
+/**
+ * Factored out common process for running test suite in python container
+ *
+ * @param pyimage Custom python image
+ */
+def test_with_container(String pyimage) {
+    withDockerServer([uri: PCIC_DOCKER]) {
+        def pytainer = docker.image(pyimage)
+
+        pytainer.inside {
+            // Installs
+            sh 'pip install -r requirements.txt -r test_requirements.txt -i http://tools.pacificclimate.org/pypiserver/ --trusted-host tools.pacificclimate.org'
+            sh 'pip install .'
+
+            // Tests
+            sh 'py.test -v --tb=short --cov --flake8'
+        }
+    }
+}
+
+
+/**
+ * Build and release package to pypi
+ */
+def push_to_pypi() {
+    withDockerServer([uri: PCIC_DOCKER]) {
+        def pytainer = docker.image('crmprtd-python35')
+
+        pytainer.inside {
+            // get twine
+            sh 'pip install twine'
+
+            // Build
+            sh 'python setup.py sdist'
+
+            // Release
+            withCredentials([usernamePassword(credentialsId: 'PCIC_PYPI_CREDS', usernameVariable: 'USERNAME', passwordVariable: 'PASSWORD')]) {
+                sh "twine upload --repository-url https://pypi.pacificclimate.org/ --skip-existing -u $USERNAME -p $PASSWORD dist/*"
+            }
+        }
+    }
+}
+
+
+node {
+    stage('Code Collection') {
+        checkout scm
+    }
+
+    stage('Testing') {
+        parallel "Python 3.5": {
+            test_with_container('crmprtd-python35')
+        },
+        "Python 3.6": {
+            test_with_container('crmprtd-python36')
+        },
+        "Python 3.7": {
+            test_with_container('crmprtd-python37')
+        }
+    }
+
+    if (BRANCH_NAME == 'master') {
```

Comment:
Do we want to push to pypi for *every* master commit, or just for tagged releases? Is there a way to check for a tag here?

Reply:
So this is something that I think will take some more work.  From what I've gathered Jenkins doesn't have a great way to do this and the alternative options I've tried have been unsuccessful.  I can take another swing at it as this is something I would want for other pipelines as well.

Reply:
Yeah, that'd be good. Thanks. What has been the challenge with it so far?

Reply:
The issue was that I could not get any information about the tags while in Jenkins.  Eventually I found that this was because the default source control checkout runs `git fetch` with the `--no-tags` flag. Once I figured that out checking for a tag was not difficult.